Only allow one group's submenu to be expanded at a time

robertknight opened the pull request:
Per feedback on Slack, change the behaviour of the groups menu so that only one submenu
item can be open at a time.

**User-facing changes**
- Only one groups menu item can be expanded at a time
- Add an animated transition to the expansion collapse of menu items and an animated transition to the background color of menu items when their submenu is expanding or collapsing. This animation avoids menu items "jumping" when expanding a submenu causes one above to collapse

Screencap: https://www.dropbox.com/s/e32441bix46bl1v/Animated%20groups%20menu.mp4?dl=0

**Internal changes**
- Add a `Slider` component which handles animating the reveal / hiding of content with an animation. Doing this correctly in a way that works across Safari/Chrome/Firefox has some details to it which mean that it isn't just a trivial CSS animation
- Move the responsibility for rendering the submenu container for the groups menu from `GroupsListItem` to `MenuItem`. The `GroupsListItem` supplies the content but `MenuItem` handles rendering it inside a container with an expand/collapse animation
- Replace the per-GroupListItem state indicating whether the item's submenu is
open with a per-GroupList state variable indicating which group (or none) has
its submenu currently expanded. This allows `GroupList` to ensure that only one menu item is open at a time

robertknight commented:
I'm currently working on adding an animation when submenus are expanded/collapsed to avoid a surprising jump when expanding a submenu results in the expanded item moving up due to a submenu above it being collapsed.

robertknight commented:
I have tweaked the animation by making the speed a little faster (150ms, I tried less but it felt too sudden to me) and animating the background color transition of the top-level menu item.

I also evaluated different ways of having the submenu content slide down:
1. Content attached to top of submenu, and more items revealed at the bottom of the submenu as it expands (as used on eg. https://www.bbc.co.uk when expanding the top navbar on mobile)
2. Content attached to bottom of submenu, with more items being revealed at the top of the submenu as it expands (as used by eg. the side navigation on https://material.io/design/ when expanding menu items)
3. Content fills current area of expanding submenu, with items "inflating" as the submenu expands (this is used in iOS)

Of the above, (1) felt subjectively best to me as it felt like it has the least superfluous motion to it.

I think I'm happy with the overall feel of it now. The tests need updating though, as I moved the responsibility for rendering the submenu from the `GroupListItem` component to the `MenuItem`
component.

robertknight commented on src/sidebar/components/menu-item.js:
```diff
@@ -58,51 +59,60 @@ function MenuItem({
-            target="_blank"
-            rel="noopener noreferrer"
+    // Wrapper element is a `<div>` rather than a `Fragment` to work around
+    // limitations of Enzyme's shallow rendering.
```
I haven't looked deeply into the details, but Enzyme complained when trying to use `wrapper.setProps()` to re-render the component that re-rendering is not supported when there are multiple top-level elements (as is the case when using `Fragment`). Using a concrete element fixes this by making it so that the output of `MenuItem` only has one top-level element.

robertknight commented:
> My main high-level feedback is curiosity about whether it would be possible to make a GroupMenuItemSubmenu or similar component, instead of passing a render callback around?

The content of the menu needs to be passed to the `MenuItem` somehow - whether that it is the `children` prop, or some other dedicated prop.

I've put together a commit that shows how the submenu could potentially be extracted into a separate component: https://github.com/hypothesis/client/commit/35de5fa725542cdcda48e9685d6859db5c422ee9. I'm not sure there is much value in this extraction myself, as it turns out that the submenu ends up needing a lot of the same derived information that is needed to render the parent menu item  (Can the user leave the group? Is there an activity URL? Can the group be selected?)

lyzadanger commented:
> I've put together a commit that shows how the submenu could potentially be extracted into a separate component: [35de5fa](https://github.com/hypothesis/client/commit/35de5fa725542cdcda48e9685d6859db5c422ee9). I'm not sure there is much value in this extraction myself, as it turns out that the submenu ends up needing a lot of the same derived information that is needed to render the parent menu item (Can the user leave the group? Is there an activity URL? Can the group be selected?)

Hmm, I see what you mean. It doesn't feel so elegant, does it? I still feel like the long-ish, anonymous render function is a little fiddly, but I don't want to grind this train to a halt. Perhaps some additional commenting to help alert future-devs, so they don't (as I did, at least) think that it's implemented this way because of some special or funky rendering conditions?

robertknight commented:
> Perhaps some additional commenting to help alert future-devs, so they don't (as I did, at least) think that it's implemented this way because of some special or funky rendering conditions?

I've simplified the API a little by replacing the `renderSubmenu` function prop with a `submenu` prop which is just the content of the submenu. The original reason for using a function was so that `MenuItem` could render the submenu only when needed, but it turned out that I wasn't using this capability.

In general I would say that having a function prop which is called to render some sub-section of a component on demand is not an uncommon pattern. A common use might be something like a list or table where the user supplies a callback to render individual items.

robertknight commented on src/sidebar/components/group-list-v2.js:
```diff
@@ -84,22 +91,32 @@ function GroupList({ serviceUrl, settings }) {
     >
       {currentGroupsSorted.length > 0 && (
         <GroupListSection
+          expandedGroup={expandedGroup}
```
That's correct, since the expanded group may in one of the other sections.

robertknight commented on src/sidebar/components/group-list-item.js:
```diff
@@ -19,19 +18,17 @@ const MenuItem = require('./menu-item');
 function GroupListItem({
   analytics,
-  defaultSubmenuOpen = false,
+  isExpanded,
```
If a component has internal state for something, adding a `default<Name>` prop is a convention for setting the initial value of that state. This is often useful in tests.

robertknight commented on src/sidebar/components/group-list-item.js:
```diff
@@ -79,23 +76,21 @@ function GroupListItem({
+      label={group.name}
+      onClick={isSelectable ? focusGroup : toggleSubmenu}
+      onToggleSubmenu={toggleSubmenu}
+      renderSubmenu={() => (
```
I've simplified this now to just having a `submenu` prop rather than `renderSubmenu`, but the original rationale was so that `MenuItem` could render the submenu only when it was wholly or partially visible.

robertknight commented on src/sidebar/components/group-list-item.js:
```diff
@@ -79,23 +76,21 @@ function GroupListItem({
+      isSelected={isSelected}
+      isSubmenuVisible={isExpanded}
+      label={group.name}
+      onClick={isSelectable ? focusGroup : toggleSubmenu}
```
Your understanding is correct. I can't immediately think of an obvious way to simplify this.

robertknight commented on src/sidebar/components/menu.js:
```diff
@@ -50,11 +50,21 @@ function Menu({
   contentClass,
   defaultOpen = false,
   label,
+  onOpenChanged,
```
I called the prop `onOpenChanged` to match `defaultOpen` which indicates whether the menu is initially open (this is mainly used in tests).

robertknight commented on src/sidebar/components/menu.js:
```diff
@@ -50,11 +50,21 @@ function Menu({
+  // Notify parent when menu is opened or closed.
+  const wasOpen = useRef(isOpen);
+  useEffect(() => {
+    if (typeof onOpenChanged === 'function' && wasOpen.current !== isOpen) {
```
Testing truthiness would be sufficient. If a non-function is passed that should be caught by prop-types checking.

robertknight commented on src/sidebar/components/group-list-item.js:
```diff
@@ -79,15 +76,15 @@ function GroupListItem({
         iconAlt={orgName(group)}
         isDisabled={!isSelectable}
-        isExpanded={isExpanded}
+        isExpanded={hasActionMenu ? isExpanded : undefined}
```
My understanding is that passing `undefined` is the same as not specifying the prop at all. I've changed this to `false` for obviousness.

robertknight commented:
This is ready for re-review. Other than addressing the comments, I have also changed the `Slider` implementation slightly to hide the content in the DOM when a submenu is fully collapsed. This prevents the submenu items from appearing in keyboard navigation order when the submenu is closed. I also found a more reliable and simpler way to ensure that the collapse animation happens reliably across browsers (see code comments).
